src/audio/sink.cpp

```diff
@@ -0,0 +1,265 @@
+	Expects(src.size() % bytes_per_sample == 0);
+	Expects(!src.empty());
+	const auto total = static_cast<size_t>(src.size());
+
+	// Only transfer as many bytes as the ring buffer can take,
+	// truncated to the nearest sample.
+	// Don't bother trying to write 0 samples!
+	auto count = std::min(total, this->ring_buf.WriteCapacity());
+	count -= (count % bytes_per_sample);
+	if (count == 0) return 0;
+
+	auto written_count = this->ring_buf.Write(src.first(count));
+	// Since we never write more than the ring buffer can take, and we're
+	// the only thread writing, the written count should equal the requested
+	// written count.
+	Ensures(written_count == count);
+	Ensures(written_count % bytes_per_sample == 0);
+	return written_count;
+}
+
+void Sdl_sink::Callback(gsl::span<std::byte> dest)
+{
+	Expects(0 <= dest.size());
+
+	// How many bytes do we want to pull out of the ring buffer?
+	const auto req_bytes = static_cast<size_t>(dest.size());
+
+	// Make sure anything not filled up with sound later is set to silence.
+	// This is slightly inefficient (two writes to sound-filled regions
+	// instead of one), but more elegant in failure cases.
+	std::fill(dest.begin(), dest.end(), std::byte{0});
+
+	// If we're not supposed to be playing, don't play anything.
+	if (this->state != Sink::State::playing) return;
+
+	// Let's find out how many bytes are available in total to give SDL.
+	//
+	// Note: Since we run concurrently with the decoder, which is also
+	// trying to modify the read capacity of the ringbuf (by adding
+	// things), this technically causes a race condition when we try to
+	// read `avail_samples` number of samples later.  Not to fear: the
+	// actual read capacity can only be greater than or equal to
+	// `avail_samples`, as this is the only place where we can *decrease*
+	// it.
+	const auto avail_bytes = this->ring_buf.ReadCapacity();
+
+	// Have we run out of things to feed?
+	if (avail_bytes == 0) {
+		// Is this a temporary condition, or have we genuinely played
+		// out all we can?  If the latter, we're now out too.
+		if (this->source_out) this->state = Sink::State::at_end;
+
+		// Don't even bother reading from the ring buffer.
+		return;
+	}
+
+	// Of the bytes available, how many do we need?  Send this amount to
+	// SDL.
+	auto bytes = std::min(req_bytes, avail_bytes);
+	// We should be asking for a whole number of samples.
+	assert(bytes % bytes_per_sample == 0);
+
+	auto read_bytes = this->ring_buf.Read(dest.first(bytes));
+
+	// We should have received a whole number of samples.
+	assert(read_bytes % this->bytes_per_sample == 0);
+	auto read_samples = read_bytes / this->bytes_per_sample;
+
+	this->position_sample_count += read_samples;
+}
+
+/* static */ std::vector<std::pair<int, std::string>> Sdl_sink::GetDevicesInfo()
+{
+	std::vector<std::pair<int, std::string>> list;
+
+	// The 0 in SDL_GetNumAudioDevices tells SDL we want playback devices.
+	const auto is = SDL_GetNumAudioDevices(0);
```

Comment:
could make use of this to initialise the vector size

Reply:
Yes, I think I saw this earlier and forgot to change it.  `clang-tidy` even told me to do something like this in another instance where I was `emplace_back`ing a fixed number of things into a `vector` xD

Reply:
Ah, I've just noticed that I have a check that the name returned by SDL is non-null.  I can't remember why this is, but if it's actually a possibility, that does make it somewhat harder to ensure the vector size and SDL device count line up